client/components/language-picker/modal.jsx

```diff
@@ -0,0 +1,189 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes, PureComponent } from 'react';
+import classNames from 'classnames';
+import { localize } from 'i18n-calypso';
+import { includes, map, noop, partial } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Dialog from 'components/dialog';
+import SectionNav from 'components/section-nav';
+import SectionNavTabs from 'components/section-nav/tabs';
+import SectionNavTabItem from 'components/section-nav/item';
+import Search from 'components/search';
+
+class LanguagePickerModal extends PureComponent {
+	static propTypes = {
+		onSelected: PropTypes.func,
+		onClose: PropTypes.func,
+	}
+
+	static defaultProps = {
+		onSelected: noop,
+		onClose: noop,
+	}
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			filter: 'popular',
+			search: false,
+			selectedLanguageSlug: this.props.selected,
+		};
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.selected !== this.state.selectedLanguageSlug ) {
+			this.setState( {
+				selectedLanguageSlug: nextProps.selected
+			} );
+		}
+	}
+
+	getFilterLabel( filter ) {
+		const { translate } = this.props;
+
+		switch ( filter ) {
+			case 'popular':
+				return translate( 'Popular languages', { textOnly: true } );
+			default:
+				return translate( 'All languages', { textOnly: true } );
+		}
+	}
+
+	getDisplayedLanguages() {
+		const { languages } = this.props;
+		const { search, filter } = this.state;
+
+		if ( search ) {
+			const searchString = search.toLowerCase();
+			return languages.filter( language => {
+				return includes( language.name.toLowerCase(), searchString );
+			} );
+		}
+
+		switch ( filter ) {
+			case 'popular':
+				const popularLanguages = languages.filter( language => language.popular );
+				popularLanguages.sort( ( a, b ) => a.popular - b.popular );
+				return popularLanguages;
+			default:
+				return languages;
+		}
+	}
+
+	handleSearch = ( search ) => {
+		this.setState( { search } );
+	}
+
+	handleClick = ( selectedLanguageSlug ) => {
+		this.setState( { selectedLanguageSlug } );
+	}
+
+	handleSelectLanguage = () => {
+		const langSlug = this.state.selectedLanguageSlug;
+		this.props.onSelected( langSlug );
+		this.handleClose();
+	}
+
+	handleClose = () => {
+		this.props.onClose();
+	}
+
+	renderTabItems() {
+		const tabs = [ 'popular', '' ];
+
+		return map( tabs, filter => {
```

Comment:
The `getFilterLabel` method is also used when rendering the `<SectionNav>` `selectedText` attribute. So it's not so easy to drop it.

There will be more tabs in this list when we implement the grouping of languages by territory (Europe, Africa, ...) We can do some refactoring at that time.